### ISSUE TITLE
Should we specify what error class is returned when a procedure encounters an error of a particular type

### ISSUE THREAD

**Issue by abouteiller** [mpi-forum/mpi-issues#691]
# Problem

We almost never specify what error class is produced by error scenarios. For example, if a message is not delivered, MPI will produce an error code of any given class that the implementation feels is relevant (including ERR_UNKNOWN).

The exception to that blanket statement is wrong argument types, for which we usually have a defined error class that represents "passing the wrong handle type/NULL handle to the procedure argument".

# Proposal

Review if, for MPI-5, we want to be more specific about what error classes are produced in what cases. The Forum is not of one voice on this, so the first task is an open discussion about the need for such a feature. 

# Changes to the Text

All procedure to list common error cases, and the expected error class produced when that error occurs. 

# Impact on Implementations

Implementation will have less freedom in choosing what class of error to produce.

# Impact on Users

Users can check for particular error classes to react to specific error behavior.

# References and Pull Requests

**Comment by cniethammer** [comment]
I think clearly specifying error classes to be returned for procedures is a good thing.

A use case I just came along: Checking if a file exists with `MPI_File_open(..., MPI_MODE_CREATE|MPI_MODE_EXCL, ...)`.
I cannot decide just by checking for MPI_SUCCESS and require implementation to return, e.g., `MPI_ERR_FILE_EXISTS`.

**Comment by abouteiller** [comment]
@qkoziol and @jeffhammond you were vocal about this issue (pro and con). Feel free to voice a better case for/against it.

**Comment by jeffhammond** [comment]
being fully explicit about errors for all functions is going to require a huge volume of text.  we should attempt to create a chapter on errors that addresses this issue without having to add text to the definition of every function.

**Comment by cniethammer** [comment]
I agree, documenting for all procedures the returned errors will be a bigger effort. But I think it would improve MPI.

I think return values should go into the bindings in some kind, e.g., as an enum/list of possible return values. This would make it accessible to tools, later.

**Comment by abouteiller** [comment]
If we can write the error code descriptions (in more details than currently) in a way that is sufficiently generic that their description can be blanket used, maybe a table with what procedure, which error classes, would suffice in the annex.

However, I feel that it is probable that what a particular error code means (and what is the standardized post-error state) may vary depending on what procedure/operation raised it, which would then have to be described within the binding itself.


One point to note is that we presumably will not be able to be exhaustive in what failure scenarios we cover. We can have error classes standardized for common expected error cases (as an extension of what we do for invalid NULL handles passed-in, or what we already do for ALLOC_MEM), but it is expected that we can't predict all the ways things can go wrong on all systems everywhere, everywhen, and implementation will have to produce other error codes, including MPI_ERR_UNKNOWN and MPI_ERR_OTHER, or other codes as deemed appropriate by the implementor for the particular failure mode.
